### PR TITLE
Add playlists and playlist items

### PR DESCRIPTION
**Pull request by robbevp**
This PR adds a basic implementation of Playlists and exposes this through the API.

The concept of a Playlist is currently the following:
* A playlist is a collection of either albums, artists or tracks. All items within one playlist should be of the same type.
  This allows us to create playlist like 'My favorite songs', 'My top 10 albums of 2022', 'All artists playing at festival X'
* The items in a playlist are ordered (though a user can choose to not use this).
* A playlist is either personal or shared. A personal playlist can only be updated or destroyed by the user it belongs to, a shared playlist can be updated all the users.
  My primary reasoning for this distinction is for subjective and objective lists. 'Robbe's top 10 ambient albums' would be a personal list. 'All artists playing at Rewire 2022' is an objective list that everyone can add to. Another option for a shared list could be 'Music for our road trip'
* A playlist has a name and a description. This allows the users to keep the name short 'Rewire 2022' and provide extra info in the description 'All artists playing at Rewire 2022'
  
### Private playlists?  
~~I was doubting to allow a user to make their personal playlists private. Currently all users can see all playlists, where a private playlist would only be visible to the user the list belongs to. I didn't think I'd ever mark a playlist private, but if there is a need for it, I wouldn't mind it either.~~

Private playlists are now added

### `PlaylistItem` needed?
We currently store the playlist items in a separate model/table. Mostly so we can use Rails' builtins for associations, and store order as an explicit field.

However, at the moment our only interaction with these items is through the `item_ids` getter/setter, so we don't necessarily need a separate model / table, but also might store `item_ids` in a json/jsonb type column on the playlists table. This would also allow us to remove the checks on the playlist's type and the item's item_type

### CODE REVIEW

**Comment by robbevp** [comment]
I'm keeping this a draft, since the controller implementation might still change a bit (especially around creating/updating lists), while working on a matching PR in the web release. However, I do think that this is ready to be looked at

File: app/policies/playlist_policy.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      scope.where(private: false).or(scope.where(private: true, user_id: user.id))
+    end
+  end
+
+  def index?
+    user.present?
+  end
+
+  def show?
+    user.present?
```
Comment (chvp) [inline app/policies/playlist_policy.rb:13]: Needs to also take into account the `private` flag, no?

File: app/models/playlist.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#  description   :string
+#  name          :string           not null
+#  playlist_type :integer          not null
+#  private       :boolean          default(FALSE)
+#  created_at    :datetime         not null
+#  updated_at    :datetime         not null
+#  user_id       :bigint
+#
+class Playlist < ApplicationRecord
+  belongs_to :user, optional: true
```
Comment (chvp) [inline app/models/playlist.rb:15]: Always storing the creator and keeping the three levels of access (private/personal/public) in an enum makes more sense to me, and it feels like it would be harder to mess up access with that system.

File: app/serializers/playlist_serializer.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+#  description   :string
+#  name          :string           not null
+#  playlist_type :integer          not null
+#  private       :boolean          default(FALSE)
+#  created_at    :datetime         not null
+#  updated_at    :datetime         not null
+#  user_id       :bigint
+#
+class PlaylistSerializer < ActiveModel::Serializer
+  attributes :id, :name, :description, :user_id, :playlist_type, :created_at, :updated_at, :item_ids, :private
```
Comment (chvp) [inline app/serializers/playlist_serializer.rb:15]: Have you tested what happens with a huge amount of `item_ids`?

**Review by chvp — CHANGES_REQUESTED** [review]

File: app/models/playlist.rb (hunk 2 above)
Comment (robbevp) [reply app/models/playlist.rb]: I've implemented this, i agree that it makes things a lot simpler. I went for `secret` rather than `private` since an enum can't define the method private (I collides with a method from ActiveRecord).

One extra point here: The user_id now only gets set add creation. This means that if user X makes playlist and then user Y makes it personal or secret, it still belongs to user X. The other option would have been to always set the user_id to the last user who edited the playlist.
I don't think this is a real world problem (since shared playlist already require the trust that another user would not just delete all the items), but wanted to point it out as an edge case

File: app/serializers/playlist_serializer.rb (hunk 3 above)
Comment (robbevp) [reply app/serializers/playlist_serializer.rb]: I just did a few test. There was some N+1 querying going on, but this is now solved.

When we have multiple playlists with a total of 60000 items between them, the request takes between 800ms and 1000ms. This is certainly something to watch out for, if a request would need to return the full 500 lists, all with a >1000 items in each one, this request would probably be painfully slow.

Maybe we need to use a lower number of playlists per page?

Comment (robbevp) [reply app/serializers/playlist_serializer.rb]: I added an extra commit, which improves this situation A LOT. The same request with 60000 playlist_items now takes 170 - 300ms and should be able to scale without much issues. It does require more (postgres-only) SQL to be present in the model, but I'd say that is worth it here?

File: app/models/playlist.rb (hunk 2 above)
Comment (chvp) [reply app/models/playlist.rb]: Is access an attribute we want to be able to change after creation? (I don't immediately see a use-case.)

Comment (robbevp) [reply app/models/playlist.rb]: I'd say so, mostly because someone might change their mind on something being personal/secret (or just create a playlist a select the wrong access by accident)

File: app/models/artist.rb [hunk 4]
```diff
@@ -42,6 +44,11 @@ def merge(other)
       ta.update(artist_id: id)
     end
 
+    # rubocop:disable Rails/SkipsModelValidations
+    # Since we only update the item_id, there aren't any validations that could fail
+    other.playlist_items.update_all(item_id: id)
+    # rubocop:enable Rails/SkipsModelValidations
+
```
Comment (chvp) [inline app/models/artist.rb:51]: Doesn't this mean there could be duplicates in a playlist? I guess that's not a failed validation, but is this something we want in this case?

File: app/models/playlist.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+#  created_at    :datetime         not null
+#  updated_at    :datetime         not null
+#  user_id       :bigint           not null
+#
+class Playlist < ApplicationRecord
+  belongs_to :user
+  has_many :items, class_name: 'PlaylistItem', dependent: :destroy
+
+  enum access: { shared: 0, personal: 1, secret: 2 }
+  enum playlist_type: { album: 1, artist: 2, track: 3 }
```
Comment (chvp) [inline app/models/playlist.rb:19]: Why not start at 0?

**Review by chvp — APPROVED** [review]

**Comment by chvp** [comment]
(Pushed a rebase since this branch didn't include the flake yet.)

File: app/models/playlist.rb (hunk 5 above)
Comment (robbevp) [reply app/models/playlist.rb:19]: Can't remember - I'll change this

File: app/models/artist.rb (hunk 4 above)
Comment (robbevp) [reply app/models/artist.rb:51]: we validate the uniqueness of an item_id within a playlist (both with validations and with an index) - but I'm not sure what would happen here. Probably an error? 

I'll check + add a test to see that the merging will always succeed (and do the same for tracks)

File: flake.nix [hunk 6]
```diff
@@ -58,7 +58,7 @@
             env = [
```
Comment (robbevp) [inline flake.nix:58]: @chvp I think these changes accidentally slipped into this PR when rebasing?

Comment (chvp) [reply flake.nix:58]: Right, I was going to make this a separate PR and then I forgot.

**Comment by robbevp** [comment]
Updated the merging strategy of tracks and artists to make sure this works when they are in the same playlist. @chvp Maybe take a short look at the changes

**Review by chvp — APPROVED** [review]

**Comment by robbevp** [comment]
Sorry - one more change, since the current result of `with_item_ids` was causing errors in the web interface

File: app/models/playlist.rb [hunk 7]
```diff
@@ -18,15 +18,15 @@ class Playlist < ApplicationRecord
   enum access: { shared: 0, personal: 1, secret: 2 }
   enum playlist_type: { album: 0, artist: 1, track: 2 }
 
-  scope :with_item_ids, -> { left_joins(:items).select('playlists.*', 'array_agg(playlist_items.item_id ORDER BY playlist_items.order ASC) as item_ids').group(:id) }
+  scope :with_item_ids, -> { left_joins(:items).select('playlists.*', 'array_remove(array_agg(playlist_items.item_id ORDER BY playlist_items.order ASC), NULL) as item_ids').group(:id) }
```
Comment (chvp) [inline app/models/playlist.rb:21]: Can you add a comment in the code on why the `array_remove(..., NULL)` is necessary? It isn't immediately obvious to me now, so it definitely won't be in the future.

File: app/models/playlist.rb [hunk 8]
```diff
@@ -18,15 +18,15 @@ class Playlist < ApplicationRecord
...
 
   validates :name, presence: true
 
   before_save :normalize_item_order
 
   def item_ids
     # If with_item_ids was used, we use the output from the SQL query
-    return read_attribute(:item_ids) if has_attribute?(:item_ids)
+    return read_attribute(:item_ids) || [] if has_attribute?(:item_ids)
```
Comment (chvp) [inline app/models/playlist.rb:29]: Is there really no way to make the SQL query return an empty array?

**Review by chvp — CHANGES_REQUESTED** [review]

File: app/models/playlist.rb (hunk 8 above)
Comment (robbevp) [reply app/models/playlist.rb:29]: We could do this with the following, but lose some readability:
```sql
COALESCE(array_remove(array_agg(playlist_items.item_id ORDER BY playlist_items.order ASC), NULL), [])
```

Comment (chvp) [reply app/models/playlist.rb:29]: I think I prefer that solution. The SQL is already complicated enough that it needs some comments IMO, so adding some slight extra complexity there seems preferable to me to adding it here.

**Comment by robbevp** [comment]
@chvp my last commit only contains some reformatting - both before and after should pass our rubocop config, but i felt that a multiline lamdba made more sense here. Feel free to drop this commit if you disagree

**Review by chvp — APPROVED** [review]
